google/cloud/security/common/gcp_api/_base_repository.py

```diff
@@ -0,0 +1,698 @@
+
+    def execute_command(self, verb, verb_arguments):
+        """Executes command (ex. add) via a dedicated http object.
+
+        GCP APIs may take minutes to complete. Therefore, callers are
+        encouraged to leverage concurrent.futures (or similar) to place long
+        running commands on a separate threads.
+
+        Args:
+          verb (str): Method to execute on the component (ex. get, list).
+          verb_arguments (dict): key-value pairs to be passed to
+              _build_request.
+
+        Returns:
+          dict: An async operation Service Response.
+        """
+        request = self._build_request(verb, verb_arguments)
+        request_submission_status = self._execute(request)
+
+        return request_submission_status
+
+    def execute_paged_query(self, verb, verb_arguments):
+        """Executes query (ex. list) via a dedicated http object.
+
+        Args:
+          verb (str): Method to execute on the component (ex. get, list).
+          verb_arguments (dict): key-value pairs to be passed to _BuildRequest.
+
+        Yields:
+          dict: Service Response.
+
+        Raises:
+          PaginationNotSupportedError: When an API does not support paging.
+        """
+        if not self._request_supports_pagination(verb=verb):
+            raise api_errors.PaginationNotSupportedError(
+                '{} does not support pagination')
+
+        request = self._build_request(verb, verb_arguments)
+
+        number_of_pages_processed = 0
+        while request is not None:
+            response = self._execute(request)
+            number_of_pages_processed += 1
+            LOGGER.debug('Executing paged request #%s',
+                         number_of_pages_processed)
+            request = self._build_next_request(verb, request, response)
+            yield response
+
+    def execute_search_query(self, verb, verb_arguments):
+        """Executes query (ex. search) via a dedicated http object.
+
+        Args:
+          verb (str): Method to execute on the component (ex. search).
+          verb_arguments (dict): key-value pairs to be passed to _BuildRequest.
+
+        Yields:
+          dict: Service Response.
+        """
+        # Implementation of search does not follow the standard API pattern.
+        # Fields need to be in the body rather than sent seperately.
+        next_page_token = None
+        number_of_pages_processed = 0
+        while True:
+            req_body = verb_arguments.get('body', dict())
+            if next_page_token:
+                req_body['pageToken'] = next_page_token
+            request = self._build_request(verb, verb_arguments)
+            response = self._execute(request)
+            number_of_pages_processed += 1
+            LOGGER.debug('Executing paged request #%s',
+                         number_of_pages_processed)
+            next_page_token = response.get('nextPageToken')
+            yield response
```

Comment:
We should probably yield response['items'] - items being the item key. This implies, that the item key needs to be passed in.

Reply:
I like keeping these as close to the raw API response as possible, and letting the higher level implementations handle extracting the data from the response.